### PR TITLE
VS2017 vcpkg support.

### PR DESCRIPTION
**Pull request by jeffhostetler**
Use the new vcpkg package manager to build (fresh) third party libraries
rather than NuGet packages when building with MSVC.  The vcpkg setup
can run with either a Command Prompt or a SDK bash window.

I have built git using VS2017 (using "make MSVC=1") and all seems well.
I have not tried it with VS2015.

@dscho  I left a TODO note in the README about updating the PERL code to
generate the .sln/.vcxproj files.  It should be relatively easy, but PERL is not my
thing.

### CODE REVIEW

**Comment by jeffhostetler** [comment]
I should have mentioned that the 3 or 4 vcpkg_*.bat setup scripts only need
to be run once in the repo.  Then the Makefile will just use the existing installation.
I did not want to try to automatically download and build vcpkg, cmake, and all the
third-party libraries as a subordinate task implicitly referenced by the Makefile.

File: compat/vcbuild/vcpkg_defines.bat [hunk 1]
```diff
@@ -0,0 +1,41 @@
+@ECHO OFF
+REM ================================================================
+REM This script builds a set of MAKEFILE definitions to allow a
+REM normal "make MSVC=1 vcpkg" to find the vcpkg-version of the
```
Comment (jeffhostetler) [inline compat/vcbuild/vcpkg_defines.bat]: This should just be "make MSVC=1".  I wrote this comment when I was trying to keep both types of targets.

**Comment by jeffhostetler** [comment]
TODO It turns out that when you build with "make MSVC=1 DEBUG=1", that the zlib .lib/.dll are named zlibd rather than zlib.  I need to update the bat file that installs them to remove the 'd' -- or to fix the ld step to know about the debug flag.

**Comment by dscho** [comment]
Awesome! I'll have a look next week.

**Comment by dscho** [comment]
> I'll have a look next week.

Or six weeks later...

Sorry!

Anyway, I gave this a look-over, and also worked on using the vcpkg stuff from inside Visual Studio. I *think* I succeeded... Could you have a look @jeffhostetler?

**Comment by dscho** [comment]
Okay, I worked a bit more on this, and now it even copies the `.dll` and `.pdb` files as intended. I had minor problems running the test suite, but those might be unrelated problems, as I ran the test suite in a plain Git for Windows Bash (i.e. not the SDK).

Sadly, a quick `git clean -dfx` made all of my records go away, so I do not even remember which tests those were ;-)

Anyway, could you have another look, @jeffhostetler?

**Comment by jeffhostetler** [comment]
@dscho I didn't see what you did to the installer for the DLLs and PDBs, but I pushed a commit to copy the 3rd-party DLLs and PDBs during the install.  I just tried a "make MSVC=1 HOME=/d/foo install" and was able to use the resulting installation in /d/foo/bin, so I think we're good.

Building from the SDK/bash worked fine and automatically fetched vcpkg, cmake, and the libraries, so I think were good there.

I did not try the .sln portion this morning.

**Comment by dscho** [comment]
So... I ran the test suite and it turns out that t6500 is failing.

Today, I found some time to investigate a bit further and it would appear as if Git uses `fscanf()` wrong: it parses a string of *up to* 256 characters via `%250c`. That is incorrect, because that says "precisely 256 characters", and not NUL terminated.

Funnily enough, further digging brings up afe2fab72c3151d02314ab5ef3708cd203c499e5... which is a commit that made it into v2.14.3, but this here branch is based on an earlier commit. So I simply backported the fix.

IIRC there was one other test failure, I am re-running the entire test suite now.

**Comment by dscho** [comment]
I added another patch that was required here to let t7814 pass; it triggered a thread-related bug in `make_environment_block()`, and I *think* it is the correct way to fix it. @jeffhostetler, could you please have a look?

**Comment by dscho** [comment]
BTW there is *one* more thing I would love to do before merging this PR: the `<Copy>` task in the VS projects interferes between projects built in parallel, because they all want to copy `.dll` files either into the top-level directory or into `t/helper/`. I would love to find a way to make that conditional on the timestamps. (Later, though, I need to take a break now.)

**Comment by PhilipOakley** [comment]
An Aside: What is the minimum needed workflow when installing the VS2017 community edition (15.5.2). 

I'm guessing that the first offered option of  the Universal Windows Development Platform" will give every thing and the kitchen sink...

I'm more thinking that "Desktop Development with C++" is the more appropriate (better /minimal) option for the GfW usage (this is for a W7 64 bit machine)

Any thoughts?

**Comment by dscho** [comment]
I pushed yet another commit, this one fixing a bug in one of my old fixups. Sorry for the breakage!

**Comment by dscho** [comment]
> What is the minimum needed workflow when installing the VS2017 community edition (15.5.2).

My intention is that you should be able to build the solution, and it does everything you need to allow running the tests in a Git Bash (from your regular Git for Windows installation, no Git for Windows SDK required).



> I'm guessing that the first offered option of the Universal Windows Development Platform" will give every thing and the kitchen sink...

It should. It also should successfully install the "vcpkg" system into a subdirectory, and build the necessary dependencies there.

> I'm more thinking that "Desktop Development with C++" is the more appropriate (better /minimal) option for the GfW usage (this is for a W7 64 bit machine)

Possibly. I don't know, quite honestly. If you could try it, and if you can then verify that it eases the initial install workload noticeably, I think that would make for a fine contribution to the wiki page.

File: compat/mingw.c [hunk 2]
```diff
@@ -1311,13 +1311,42 @@ static char *path_lookup(const char *cmd, int exe_only)
 }
 
 #if defined(_MSC_VER)
+
+/* We need a stable sort */
+#ifndef INTERNAL_QSORT
+#include "qsort.c"
+#endif
+
+/* Compare only keys */
```
Comment (jeffhostetler) [inline compat/mingw.c]: the previous version used case-insensitive routines.  the new version is doing exact match.  do we care?

File: compat/mingw.c [hunk 3]
```diff
@@ -1328,162 +1357,96 @@ static wchar_t *make_environment_block(char **deltaenv)
...
 	 */
-	if (deltaenv && deltaenv[0] && *deltaenv[0]) {
-		for (k = 0; deltaenv && deltaenv[k] && *deltaenv[k]; k++) {
-			if (strchr(deltaenv[k], '=') == NULL)
-				nr_delta_del++;
-			else {
-				maxlen += strlen(deltaenv[k]) + 1;
-				nr_delta_ins++;
-			}
+	if (!deltaenv || !*deltaenv) {
```
Comment (jeffhostetler) [inline compat/mingw.c]: If there are no deltaenv values, we should be able to just return the value from GetEnvironmentStringW() -- or is there an issue of which free() should be used to free it?

File: compat/mingw.c [hunk 4]
```diff
@@ -1328,162 +1357,96 @@ static wchar_t *make_environment_block(char **deltaenv)
...
 	}
-	maxlen++;
 
 	/*
-	 * Allocate blocks for inserted and deleted items.
-	 * The individual pointers in the 'C' arrays will point into here.
-	 * We will use the wblock_ins as the final result.
+	 * If there is a deltaenv, let's accumulate all keys into `array`,
+	 * sort them using the stable git_qsort() and then copy, skipping
+	 * duplicate keys
```
Comment (jeffhostetler) [inline compat/mingw.c]: Do we care about order within the deltaenv?  That is, if they pass a delete-then-set vs a set-then-delete, do we want different results?

File: compat/mingw.c [hunk 5]
```diff
@@ -1328,162 +1357,96 @@ static wchar_t *make_environment_block(char **deltaenv)
...
 
-		/* lookup v_j in deltaenv portion of result set */
-		for (k_ins = 0; k_ins < nr_delta_ins; k_ins++) {
-			if (wcsnicmp(v_j, wptrs_ins[k_ins], len_j_eq) == 0)
-				goto skip_it;
-		}
+	/* convert the deltaenv, appending to array */
+	for (i = 0, p3 = wdeltaenv; deltaenv[i]; i++) {
+		size_t s = strlen(deltaenv[i]) + 1, wlen;
+		wlen = xutftowcs(p3, deltaenv[i], s * 2);
```
Comment (jeffhostetler) [inline compat/mingw.c]: s/2/sizeof(wchar_t)/ ??

File: compat/mingw.c (hunk 3 above)
Comment (dscho) [reply compat/mingw.c]: Yep, the function to use is called `FreeEnvironmentStringsW()`... not so easily integrated into Git's workflow. Hence copy and throw over the wall...

File: compat/mingw.c (hunk 4 above)
Comment (dscho) [reply compat/mingw.c]: Yes, that's why I need the stable sort, and that's why `wenvcmp()` does not simply `wcscmp()` (or `_wcsicmp()`).

File: compat/mingw.c (hunk 2 above)
Comment (dscho) [reply compat/mingw.c]: Hrm. You're right...

Comment (dscho) [reply compat/mingw.c]: ... and there seems to not be any way to compare `WCHAR`s case-insensitively. I guess there are issues with compound characters or some such.

So how about this instead?

```c
size_t p_len, q_len;
int ret;

/* Find end of keys */
for (p_len = 0; p[p_len] && p[p_len] != L'='; p_len++)
    ; /* do nothing */
for (q_len = 0; q[q_len] && q[q_len] != L'='; q_len++)
    ; /* do nothing */

/* Are keys identical (modulo case)? */
if (p_len == q_len && !_wcsnicmp(p, q, p_len))
    return 0;

ret = _wcsnicmp(p, q, p_len < q_len ? p_len : q_len);
return ret ? ret : (p_len < q_len ? -1 : +1);
```

File: compat/mingw.c (hunk 5 above)
Comment (dscho) [reply compat/mingw.c]: (And the `* 2` says: any UTF-8 character can be converted to at most two wide characters)

**Comment by jeffhostetler** [comment]
@PhilipOakley I think you should be able to use the free edition of VS2017.  the Universal Windows Platform refers to something else IIRC.  You should install the C/C++ compiler option (I forget what it's called in the new installer).  By default I think the installer only defaults to the managed code products.

Any of the VS2017 versions Update 2 or later should work the best.  Update 2 added the VsWhere command that we use to find the compiler and linker (necessary when using a terminal that doesn't inherit its environment from a VCVARS-based command prompt).

I have built it using the new VCPKG scripts using VS2015 and earlier versions of VS2017, but it has been a while since I tried.  And I've not tried the new versions of the .sln/.vcxproj files with VS2015 -- only from the SDK command line.

Let us know if you have questions or problems getting it to build.

File: compat/mingw.c (hunk 2 above)
Comment (jeffhostetler) [reply compat/mingw.c]: I can't tell in this GH diff viewer, so I'll ask it here.  The deltaenv[] contains "KEY\\0" (and possibly "KEY=\\0") for items to be deleted from the existing settings.  Have we already converted them to canonical form L"KEY=\\0" by the time we get here?

File: compat/mingw.c (hunk 4 above)
Comment (jeffhostetler) [reply compat/mingw.c]: I mean if they pass in deltaenv:

    KEY
    KEY=foo

vs

    KEY=foo
    KEY

I would expect different results when de-duping deltaenv.  In the first case, the result has KEY=foo, but in the second case the result does not have KEY, right?

I'm not seeing how the wenvcmp() preserves this, but i'll look again.

Comment (dscho) [reply compat/mingw.c]: Yes, that is my reading about the order, too.

So how does the stable sort together with the complicated way `wenvcmp()` ensures to compare *only* the keys do that?

Well, first of all, `wenvcmp()` tries to match only the keys, i.e. `KEY=foo` is identical to `KEY`. If we used something like `wcscmp()` directly, `KEY` would always sort *before* `KEY=foo`. Not what we want in your second example.

Second, the stable sort ensures that keys considered identical by the comparison function preserve their relative order. That is, if you add `KEY=foo`, then a couple of other variables, and then `KEY`, and if your comparison function considers `KEY=foo` to be equivalent to `KEY`, then the sorted output will still have `KEY=foo` before `KEY`, still.

The [final copying loop](https://github.com/git-for-windows/git/pull/1304/files/5a28d2e6c4742be4186e361392388f89d5c83e3c..1e902ee76e6ca57516e3a15f2f55a4ac911f28bd#diff-a0c7aa297d8da1149ce761c31dc59328R1423) then makes sure that entries are skipped that are to-be-deleted entries, and also entries are skipped that are overridden by the respective next entry (for which I still need to convert the case-sensitive comparison, too).

**Comment by dscho** [comment]
I pushed some more commits, one that has been a *looong* standing bug where `git-receive-pack.exe` was not generated and hence the wrong .exe was tested.

Also, the comparison in `wenvcmp()` is now case-insensitive again, as you suggested.

File: compat/mingw.c [hunk 6]
```diff
@@ -1328,162 +1357,96 @@ static wchar_t *make_environment_block(char **deltaenv)
...
+			continue;
+
+		p = array[i];
+
+		/* Skip any duplicate */
+		if (i + 1 < nr) {
+			wchar_t *next = array[i + 1];
+			size_t n = equal - p;
+
+			if (!wcsncmp(p, next, n) && (!next[n] || next[n] == L'='))
```
Comment (jeffhostetler) [inline compat/mingw.c]: @dscho  Shouldn't this be case insensitive ??

File: compat/mingw.c (hunk 2 above)
Comment (whoisj) [reply compat/mingw.c]: @dscho can [CompareStringEx](https://msdn.microsoft.com/en-us/library/windows/desktop/dd317761(v=vs.85).aspx) get mapped into GfW? It's kind of like `_wcsnicmp` but has extended option which might allow you to get exactly what you need.

It's Win32 API so it lacks conformance to the style of GfW, but the functionality win might be worth it.

**Comment by dscho** [comment]
And the test suite passed!

File: compat/mingw.c (hunk 6 above)
Comment (dscho) [reply compat/mingw.c]: I fixed this together with the `wenvcmp()` change to make things case-insensitive...

File: compat/mingw.c (hunk 2 above)
Comment (dscho) [reply compat/mingw.c]: Sorry, I missed this before merging the PR... I had a quick look and was *overwhelmed* with the amount of options to compare case-insensitively. I never knew how many options there are!

However, I have the impression that Win32 API's environment handling considers environment keys as identical if `_wcsicmp()` considers them identical. Am I correct?

**Pull request merged**